### ISSUE TITLE
Crash: Linter error: PHPStan crashed in the previous run probably because of excessive memory consumption

### ISSUE THREAD

**Issue by sbrl** [BradCrumb/atom-linter-phpstan#9]
Spoiler alert: It didn't use all that much memory.

Hey, @BradCrumb! I found another bug:

```
[Linter] Error running PHPStan Error: PHPStan crashed in the previous run probably because of excessive memory consumption.
It consumed around 10 MB of memory.


To avoid this issue, allow to use more memory with the --memory-limit option.
    at ChildProcess.<anonymous> (/home/sbrl/.atom/pac…xec/lib/index.js:56)
    at emitTwo (events.js:126)
    at ChildProcess.emit (events.js:214)
    at maybeClose (internal/child_process.js:925)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:209)
```

For reference, the default limit is 100M, and tried it with 1024M and it still didn't work (it didn't use anywhere _near_ that amount of memory).

It works absolutely fine on the command line:

```
$ ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse core/02-environment.php  --level 0
Note: Using configuration file some/path/to/peppermint/phpstan.neon.
 1/1 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                
 [OK] No errors                                                       
```

This one occurs every time it lints on save.

More info available on request of course.

**Comment by mallardduck** [comment]
Hey @sbrl,
It might be helpful to run PHPstan from a CLI with all the same flags. Maybe I'm wrong, but it seems like the other flags could have some affect. Depending on how you config the linter here is the relevant code to figure our which flags are used.

                const parameters = [
                    'analyze',
                    '--error-format',
                    'raw',
                    '--no-progress',
                    filePath
                ];

                // PHPStan levels
                const levels = [
                    '0', '1', '2', '3', '4', '5', '6', '7', 'max'
                ];

                if (levels.indexOf(this.level) != -1) {
                    parameters.push('-l', this.level);
                }

                if (autoloadFile) {
                    parameters.push('-a', autoloadFile);
                }

                if (this.memoryLimit) {
                    parameters.push('--memory-limit', this.memoryLimit);
                }

                if (confFile) {
                    parameters.push('-c', confFile);
                }

Seems like you did do some tests like this - but only included the level. I'm wondering if some combination of the other flags affect how much memory is used?

**Comment by sbrl** [comment]
I see, @mallardduck. How can I get it to tell me the exact flags it used though? Does it log this to the console or something, or would I have to use a tool like [execsnoop](https://unix.stackexchange.com/a/345021/64687) to peel it out manually?

**Comment by mallardduck** [comment]
In theory you could do something like that - however with the code sample I included you can manually 'unroll' the flags that are used. By default you'll always have the following:

    ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress ${the_file} -l ${the_level}

Then based on your configuration of the plugin, you may also have Autoload, memory limit and config flags. Using the line above you can replace the variables I included with your actual values. Then add those additional flags as needed.

**Comment by sbrl** [comment]
Thanks, @mallardduck! Here's a variety of runs:

```
$ ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress modules/feature-user-table.php -l 0
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:2:Function register_module not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:31:Function add_action not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:36:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:46:Call to static method render_username() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:52:Function module_exists not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:71:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:91:Function add_action not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:96:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:110:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:114:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:117:Function generate_password not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:120:Function hash_password not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:126:Function save_settings not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:128:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:132:Function email_user not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:134:Function full_url not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:134:Function full_url not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:162:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:189:Function add_action not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:194:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:198:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:202:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:207:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:210:Function hash_password not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:211:Function save_settings not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:213:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:216:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:242:Function add_action not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:247:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:251:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:255:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:258:Function email_user not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:272:Function save_settings not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:274:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:277:Call to static method render_main() on an unknown class page_renderer.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:281:Function add_help_section not found.
/home/sbrl/Documents/code/php/peppermint/modules/feature-user-table.php:290:Function generate_password not found while trying to analyse it - autoloading is probably not configured properly.
$ ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress modules/action-raw.php -l 0
/home/sbrl/Documents/code/php/peppermint/modules/action-raw.php:2:Function register_module not found.
/home/sbrl/Documents/code/php/peppermint/modules/action-raw.php:26:Function add_action not found.
/home/sbrl/Documents/code/php/peppermint/modules/action-raw.php:39:Function add_help_section not found.
 ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress core/02-environment.php -l 0
/home/sbrl/Documents/code/php/peppermint/core/02-environment.php:10:Undefined variable: $settings
/home/sbrl/Documents/code/php/peppermint/core/02-environment.php:24:Undefined variable: $settings
/home/sbrl/Documents/code/php/peppermint/core/02-environment.php:32:Undefined variable: $settings
/home/sbrl/Documents/code/php/peppermint/core/02-environment.php:60:Undefined variable: $settingsFilename
```

The project in question is [Pepperminty Wiki](https://github.com/sbrl/Pepperminty-Wiki), and I haven't yet looked at writing a config file to reduce the number of 'errors' I'm getting due to the custom packaging system.

**Comment by mallardduck** [comment]
What about your memory limit? If you don't have one set in the settings for this plugin you should set the flag for default. So it would be:

    ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress ${the_file} -l 0 --memory-limit 100M

**Comment by mallardduck** [comment]
After setting this flag if you see issues then your problem (and solution) is that you're only using a 100M memory limit. So if you blow that up to a higher size it should work just fine! You could even maybe set it to the same as your PHP CLI. To figure it out:

    php -i|grep mem

**Comment by sbrl** [comment]
Setting the memory limit flag in the terminal works fine too:

```
$ ~/.config/composer/vendor/phpstan/phpstan/bin/phpstan analyse --error-format raw --no-progress core/07-unpack.php -l 0 --memory-limit 100M
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:7:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:8:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:9:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:10:Function delete_recursive not found.
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:10:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:12:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:13:Undefined variable: $paths
/home/sbrl/Documents/code/php/peppermint/core/07-unpack.php:25:Undefined variable: $paths
```

Also:

```
$ php -i | grep memory_limit
memory_limit => -1 => -1
```

...it still doesn't appear to work in Atom though @mallardduck :confused:

**Comment by mallardduck** [comment]
@sbrl Good news! I have now run into this issue after realizing I was still on my own version of the package. Switching from the one I had published to this one now yields the issues you were seeing. I only, literally, just now found this issue on my own machine - so I'll do some more testing on this.

**Comment by sbrl** [comment]
Thanks :D

I await a fix then.

**Comment by mallardduck** [comment]
So I've had poor luck tracking it down since as I've toyed around it's not behaving as I expected (read: not consistent) it would. However even this gave me some ideas about what's failing potentially - but since it's not consistent it's slower going that I'd hope. 

I'll keep at it though and hopefully have some new ideas after taking a break.

----

# UPDATE:
I've been able to further replicate the issue by pushing my configurations to extreme circumstances.

For example I switched from working on some of my libraries/packages to a larger app/site project. Then I was able to reproduce the issue more consistently - taking that a step further, I decided to reproduce the issue completely outside of atom as well.

Doing this successfully removed atom as a 'lynch pin' for this issue since it's clearly just an error bubbling up as I expected.

Here are the errors I observed while reproducing the issue via CLI:

### Command:
```
{path}/vendor/bin/phpstan analyze --error-format raw --no-progress {path}/web/app/themes/site-theme/resources/functions.php -l 0 -a {path}/web/app/themes/site-theme/vendor/autoload.php --memory-limit 256M
```


### Error Run 1:
```
Note: Using configuration file {path}/phpstan.neon.
PHP Fatal error:  Allowed memory size of 104857600 bytes exhausted (tried to allocate 20480 bytes) in phar://{path}/vendor/phpstan/phpstan-shim/phpstan/vendor/nikic/php-parser/lib/PhpParser/Lexer.php on line 255

Fatal error: Allowed memory size of 104857600 bytes exhausted (tried to allocate 20480 bytes) in phar://{path}/vendor/phpstan/phpstan-shim/phpstan/vendor/nikic/php-parser/lib/PhpParser/Lexer.php on line 255
```

### Error Run 2:
```
Note: Using configuration file {path}/phpstan.neon.
PHPStan crashed in the previous run probably because of excessive memory consumption.
It consumed around 38 MB of memory.


To avoid this issue, allow to use more memory with the --memory-limit option.
```

Ultimately the - incorrect - suspicion I had here involving atom was that potentially it was failing to load the config settings and also failing to load the default values. Thus not sending any `--memory-limit` value to the command Atom executes for us.

However I consistently observed the value getting set and only saw it fail to load in when I tweaked code. So these along with the CLI replication of the exact issue make it clear this is more of an 'upstream' issue occurring with PHPStan rather than this Atom integration for PHPSTan.

Being even more specific, only after I began testing with the WordPress Stubs for PHPStan did I push PHPStan far enough to cause issues in CLI.

All that said - I think this isn't an issue we can address at this packages levels. I would suggest that you review and/or create an Issue in the [phpstan/phpstan](https://github.com/phpstan/phpstan) repo.

Hope that helps!

**Comment by mallardduck** [comment]
Hiya! Please see the update I've posted - I go much more in depth there - going to close this one out as I've ruled out atom specific issues causing this error.

**Comment by sbrl** [comment]
Hey,  thanks for the extensive testing! I'll open an issue against phpstan/phpstan that references this issue then.

**Comment by mallardduck** [comment]
Hey! Something just came to mind - the integration we have here will pretty much always pass the memory-limit parameter to phpstan. So if you don't have one set it's always limited to 100M, regardless of your PHP.ini file.

With that revelation I went back and made sure that a setting of `-1` would work with this plugin. It didn't but I promptly fixed that with #11 and #12. So there's a new version released 2.1.1 - I would highly suggest updating.

Once you update, I think you should double check that you have the integration configured with a memory limit - since it's passing the command a limit regardless of your php.ini limit. So as of those patches you should now be able to set a limit of -1 and it's parsed correctly via the PHPStan cli.

All that said - this should improve things, but I'm still seeing the error happen in atom. So I'm going to reopen this and keep an eye on the issue you opened at phpstan: https://github.com/phpstan/phpstan/issues/2694

**Comment by ondrejmirtes** [comment]
Hi, this can easily happen if you trry to run PHPStan on a broken file, for example a class that extends an unknown class.

It should be happening less on 0.12.x because we're removing the `.memory_limit` file on these crashes. Can you test it if you're able to reproduce it on the latest version?

**Comment by mallardduck** [comment]
@ondrejmirtes Thanks for dropping in here - I'm not sure what version I was using to test. I'll make sure I'm on that version and replicate it again to see how things go and report back. @sbrl if you can do the same that'd be great.

**Comment by sbrl** [comment]
Ah, thanks for that @mallardduck. Unfortunately, it doesn't appear to be working as intended. Now I get this error message:

```
/home/sbrl/.atom/packages/linter/lib/linter-registry.js:154 [Linter] Error running PHPStan Error: PHPStan crashed in the previous run probably because of excessive memory consumption.
It consumed around 31 MB of memory.


To avoid this issue, allow to use more memory with the --memory-limit option.
    at ChildProcess.<anonymous> (/home/sbrl/.atom/packages/linter-tidy/node_modules/sb-exec/lib/index.js:56)
    at ChildProcess.emit (events.js:182)
    at maybeClose (internal/child_process.js:962)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:251)
```

![Selection_041](https://user-images.githubusercontent.com/9929737/70467404-04ce6e00-1abd-11ea-92bd-5a9987374fc3.png)

![sbrl@Riikaan: ~_042](https://user-images.githubusercontent.com/9929737/70467484-36473980-1abd-11ea-85b2-5e4ae586bb9b.png)

**Comment by sbrl** [comment]
![Selection_043](https://user-images.githubusercontent.com/9929737/70467565-61318d80-1abd-11ea-85de-9ba9a449cd00.png)

**Comment by ondrejmirtes** [comment]
It would be useful to see the output of the previous run, this message is just a consequence of that.

**Comment by sbrl** [comment]
Sorry for taking so long to get back to you - I've been very busy with PhD work. What do you mean by "previous run", and where would I find that, @ondrejmirtes?

**Comment by ondrejmirtes** [comment]
It's plain English. You run PHPStan once - it crashes. You run PHPStan for the second time after that - it outputs this message. So the cause is in the first PHPStan run.

**Comment by ondrejmirtes** [comment]
It can also simply happen if you run PHPStan multiple times over each other in parallel.

**Comment by sbrl** [comment]
Sorry, I'm still not understanding what you're asking for. Note that I don't interpret things in the way most other people do (I have Asperger's Syndrome), so I've probably got confused by your explanation.

If I enable the PHPStan Atom plugin and open / edit / save a bunch of files, the following is a complete transcript from the Atom debug console: https://pastebin.com/raw/L15yzc5k

**Comment by mallardduck** [comment]
> It can also simply happen if you run PHPStan multiple times over each other in parallel.

This is essentially how the linter plugins for Atom work. Instead of batching the linting of these files, meaning getting a list of files to lint and passing them to a single phpstan instance. For every php file it wants to lint, it will spawn a process to lint that file.

Another way to look at it is like this...you can manually use PHPSTAN a few ways:

1. you point it at a folder and PHPSTAN itself finds and lints the PHP files.
2. You can also execute PHPSTAN pointing at a single file path.

Option 1 is more efficient as it only spawns a single PHPstan instance per project. Option 2, can only lint a single file at one time - unless you script something to fire off multiple PHPstan processes each linting a single file.

This second method is how Atom linting plugins work and is causing us issues with PHPSTAN in some cases. I'm not sure if there's a clear and obvious solution to the issue yet, but when I have time I will be looking into this eventually.

**Comment by sbrl** [comment]
Ah, I see! Perhaps a queuing system would help here? Then it could process them 1 at a time.

Also, if the process is long-lived perhaps there's a memory leak involved when many files have been linted?

**Comment by mallardduck** [comment]
@sbrl so I've implemented something that's technically a fix for this issue. However it has some potential downsides and might make other debugging harder. So it's not something I'm super ready to merge yet...I might feel better if I add a toggle for a "Debug Mode" that disables the patch.

Basically the issue here is out of reach for us to implement a direct solution. This is partially because the issue is _essentially_ a collision of behaviors between phpstan, atom and our dependencies that execute phpstan for atom. More details on that aspect and potential alternative fixes in the PR.

**Comment by sbrl** [comment]
Ah, thanks for the hard work on fixing this!

**Comment by mallardduck** [comment]
For sure! Once I have time to fix up the PR (to add a toggle for Debug mode) I'll get it merged in.

**Comment by mallardduck** [comment]
Whoops. Hit wrong button.

**Comment by mallardduck** [comment]
@sbrl - I've added a "debug mode" toggle as mentioned. This allowed me to merge in the PR that I previously created a fix in. So I've merged that PR in and closed the PR duplicate #16 that was opened too.

I think this issue should be in better shape with that change merged in. Especially when you consider the improvements made in PHPstan in the time since the issue was opened.
